### DISCUSSION TITLE
Can't get SWS with systemd on ArchLinux

### DISCUSSION THREAD

**Discussion by bubnenkoff** [static-web-server/static-web-server#227] [Q&A] [answered]
Hi! I have found https://github.com/static-web-server/static-web-server/tree/master/systemd
And created this 3 files.

Error log:
```
[root@320099 ~]# systemctl start static-web-server
Job for static-web-server.service failed because of unavailable resources or another system error.
See "systemctl status static-web-server.service" and "journalctl -xeu static-web-server.service" for details.
[root@320099 ~]# journalctl -xeu static-web-server.service
░░ Defined-By: systemd
░░ Support: https://lists.freedesktop.org/mailman/listinfo/systemd-devel
░░
░░ Automatic restarting of the unit static-web-server.service has been scheduled, as the result for
░░ the configured Restart= setting for the unit.
Jun 22 17:02:16 320099.simplecloud.ru systemd[1]: Stopped Static Web Server.
░░ Subject: A stop job for unit static-web-server.service has finished
░░ Defined-By: systemd
░░ Support: https://lists.freedesktop.org/mailman/listinfo/systemd-devel
░░
░░ A stop job for unit static-web-server.service has finished.
░░
░░ The job identifier is 18436 and the job result is done.
Jun 22 17:02:16 320099.simplecloud.ru systemd[1]: static-web-server.service: Failed to load a named file descriptor: No such >
Jun 22 17:02:16 320099.simplecloud.ru systemd[1]: static-web-server.service: Failed to run 'start' task: No such file or dire>
Jun 22 17:02:16 320099.simplecloud.ru systemd[1]: static-web-server.service: Failed with result 'resources'.
░░ Subject: Unit failed
░░ Defined-By: systemd
░░ Support: https://lists.freedesktop.org/mailman/listinfo/systemd-devel
░░
░░ The unit static-web-server.service has entered the 'failed' state with result 'resources'.
Jun 22 17:02:16 320099.simplecloud.ru systemd[1]: Failed to start Static Web Server.
░░ Subject: A start job for unit static-web-server.service has failed
░░ Defined-By: systemd
░░ Support: https://lists.freedesktop.org/mailman/listinfo/systemd-devel
░░
░░ A start job for unit static-web-server.service has finished with a failure.
░░
░░ The job identifier is 18436 and the job result is failed.
```

I have not ideas why it's failing? 

When it started with systemd witch folder it use as `public`? Does it take settings from:
```
SERVER_ROOT=/var/www/html
SERVER_HTTP2_TLS=true
SERVER_HTTP2_TLS_CERT=/etc/static-web-server/example_org_fullchain.pem
SERVER_HTTP2_TLS_KEY=/etc/static-web-server/example_org_privkey.pem
SERVER_LOG_LEVEL=warn
```
In case of direct run it seems that it want `./public` folder and not use folder above
```
[root@320099 default]# /usr/local/bin/static-web-server
2023-06-22T14:47:07.281192Z ERROR static_web_server::server: server failed to start up: root directory was not found or inaccessible

Caused by:
    path ./public was not found or inaccessible
```

**Comment by joseluisq** [comment]
I don't know what is your setup but what you can do for testing is use CLI options instead `EnvironmentFile` in your `static-web-server.service` and see how it goes, obviously if you have the root directory in place.

For example something like:


```ini
ExecStart=/usr/local/bin/static-web-server --fd 0 --config-file /some/optional/file --root /some/root/dir/
```

Also, check please [the docs](https://static-web-server.net/features/file-descriptor-socket-passing/#testing) for testing the service with the socket activation.

**Comment by bubnenkoff** [comment]
@joseluisq Yes, with this string all works. What does it mean? It did not find dir for serving from config? 
Should I stay config like you show or I need to ass something another to it?
```
[root@320099 system]# systemctl start static-web-server
[root@320099 system]# systemctl status static-web-server
● static-web-server.service - Static Web Server
     Loaded: loaded (/etc/systemd/system/static-web-server.service; disabled; vendor preset: disabled)
     Active: activating (auto-restart) (Result: exit-code) since Thu 2023-06-22 22:15:46 MSK; 3s ago
TriggeredBy: ● static-web-server.socket
    Process: 10929 ExecStart=/usr/local/bin/static-web-server --fd 0 --config-file /var/www/html --root /var/www/html (code=e>
   Main PID: 10929 (code=exited, status=216/GROUP)
        CPU: 3ms
```

**Reply by joseluisq** [reply, marked as answer]
First, what you are doing instead of binding SWS to a particular port, you are binding to a file descriptor (`--fd 0` ). This is great when you want to sandbox programs like SWS via `systemd` for example.
All of this is called [socket activation](http://0pointer.de/blog/projects/socket-activation.html) and you can read more about it in [the docs](https://static-web-server.net/features/file-descriptor-socket-passing/#testing).

Regarding the error. It was about SWS complaining about the default root `public` dir not being available because most probably when you used the environment `etc_default_static-web-server` file, it was not found by `systemd` when starting your SWS service, and therefore the environment variables were not loaded. That's why SWS was using the defaults like root `public` dir, etc.
What I can recommend is that if you want to use env variables then place the environment `etc_default_static-web-server` file in an accessible directory and then reference that file using its absolute path in your service like `EnvironmentFile=/etc/default/etc_default_static-web-server`.
Additionally, you can use a config file as well or CLI arguments directly if you prefer.

**Comment by bubnenkoff** [comment]
Perfect thanks! The issue was because:
`EnvironmentFile=/etc/default/static-web-server`
Should point at `default_static-web-server`. Name should be same `etc_default_static-web-server` or `default_static-web-server`
